### PR TITLE
send messages to multiple nagios endpoints

### PR DESCRIPTION
**Pull request by Bonko**
this can be e.g. useful for softmigrations to a new monitoring host

### CODE REVIEW

File: spec/classes/periodicnoise_spec.rb [hunk 1]
```diff
@@ -20,4 +20,20 @@
...
+    should contain_file('/etc/periodicnoise/config.ini')\
+      .with_content(Regexp.new Regexp.escape 'OK       = printf "somehost.example.com;%(event);0;%(message)\n" |/usr/sbin/send_nsca -H monitor.example.com -d ";"')
+  end
+
+  context 'send to multiple hosts' do
+    let (:params) {{
+      :send_as_host => 'somehost.example.com',
+      :send_to_host => ['monitor1.example.com', 'monitor2.example.com'],
+    }}
+    it 'should e configured globally to send messages to 2 hosts' do
```
Comment (PirminTapken) [inline spec/classes/periodicnoise_spec.rb]: `s/e/be/`

**Comment by Bonko** [comment]
hm..those tests are green on my machine..

**Comment by Bonko** [comment]
rechecked: tests are green with ruby1.8.7 but fail with 1.9.3

File: templates/nsca.erb [hunk 2]
```diff
@@ -1,5 +1,10 @@
 [monitoring]
-OK       = printf "<%= @send_as_host -%>;%(event);0;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-WARNING  = printf "<%= @send_as_host -%>;%(event);1;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-CRITICAL = printf "<%= @send_as_host -%>;%(event);2;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-UNKNOWN  = printf "<%= @send_as_host -%>;%(event);3;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
+<% if @send_to_host.is_a?(String) -%>
+<% hosts = [@send_to_host] -%>
+<% else -%>
+<% hosts = @send_to_host -%>
```
Comment (mlafeldt) [inline templates/nsca.erb]: The idiomatic Ruby way is to use `Array(@send_to_host)`.

Comment (mlafeldt) [inline templates/nsca.erb]: To make it more clear:

```$ pry
[1] pry(main)> Array("Bonko")
=> ["Bonko"]
[2] pry(main)> Array(["Bonko"])
=> ["Bonko"]
[3] pry(main)>
```

:)

**Comment by nightlyone** [comment]
Awesome hack! Didn't expect this to be so easy :heart_eyes:

File: spec/classes/periodicnoise_spec.rb [hunk 3]
```diff
@@ -20,4 +20,20 @@
       .with_content(%r{somehost.example.com}) \
       .with_content(%r{monitor.example.com})
   end
+
+  it 'should be configured globally to send messages to one host' do
+    should contain_file('/etc/periodicnoise/config.ini')\
+      .with_content(Regexp.new Regexp.escape 'OK       = printf "somehost.example.com;%(event);0;%(message)\n" |/usr/sbin/send_nsca -H monitor.example.com -d ";"')
```
Comment (mlafeldt) [inline spec/classes/periodicnoise_spec.rb]: Please use brackets to make this more readable. Also, doesn't `Regexp.escape` already return a regexp?

File: templates/nsca.erb [hunk 4]
```diff
@@ -1,5 +1,10 @@
 [monitoring]
-OK       = printf "<%= @send_as_host -%>;%(event);0;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-WARNING  = printf "<%= @send_as_host -%>;%(event);1;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-CRITICAL = printf "<%= @send_as_host -%>;%(event);2;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
-UNKNOWN  = printf "<%= @send_as_host -%>;%(event);3;%(message)\n" |/usr/sbin/send_nsca -H <%= @send_to_host -%> -d ";"
+<% if @send_to_host.is_a?(String) -%>
```
Comment (nightlyone) [inline templates/nsca.erb]: You might want to omit all of these statements, if `send_to_host` is empty or undefined, etc.

File: spec/classes/periodicnoise_spec.rb (hunk 3 above)
Comment (Bonko) [inline spec/classes/periodicnoise_spec.rb]: what do you mean?

Comment (mlafeldt) [inline spec/classes/periodicnoise_spec.rb]: I thought that `Regexp.escape` returns a regexp object, but it turns out it doesn't. So ignore that comment.

Please still use brackets, e.g. `Regexp.new(Regexp.escape(..))`.

**Comment by Bonko** [comment]
@mlafeldt @nightlyone incorporated your feedback I think, PTAL

**Comment by mlafeldt** [comment]
One more thing: doesn't this require a different error handling? I'm not sure what the code does at the moment, but with the changes, only the status code of the very last command (host) will be returned due to the `;` chaining.

Chaining commands with `&&` might not be ideal either as commands might be skipped.

**Comment by nightlyone** [comment]
The code does `sh -c "STUFF"` where STUFF is everything after the `=` sign until the line ending in the config. So it will be handled like a shell oneliner. see https://github.com/Jimdo/periodicnoise/blob/master/cmd/pn/monitor.go#L112-L119 for reference.

The chaining issue still persists. You can either ignore everything but the last error via `;` or make it fail all on the first failure using `&&` as @mlafeldt mentioned. I would vote for chaining via semicolon as this makes a best effort on reporting and should be the safest for transition to icinga. 

If you put the icinga host last, you will get it's errors in the log, which might be more useful for debugging it.

**Comment by Bonko** [comment]
I agree with @nightlyone  here

**Pull request merged**